pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -260,19 +268,32 @@ public HelixAdmin getHelixAdmin() {
    */
   private HelixManager registerAndConnectAsHelixParticipant() {
     HelixManager helixManager = HelixManagerFactory
-        .getZKHelixManager(_helixClusterName, CommonConstants.Helix.PREFIX_OF_CONTROLLER_INSTANCE + _instanceId,
-            InstanceType.PARTICIPANT, _helixZkURL);
+        .getZKHelixManager(_helixClusterName, _controllerParticipantInstanceId, InstanceType.PARTICIPANT, _helixZkURL);
+
+    StateMachineEngine stateMach = helixManager.getStateMachineEngine();
+    MasterSlaveStateModelFactory factory = new MasterSlaveStateModelFactory();
+    stateMach.registerStateModelFactory(MasterSlaveSMD.name, factory);
     try {
       helixManager.connect();
       return helixManager;
     } catch (Exception e) {
-      String errorMsg =
-          String.format("Exception when connecting the instance %s as Participant to Helix.", _instanceId);
+      String errorMsg = String.format("Exception when connecting the instance %s as Participant to Helix.",
+          _controllerParticipantInstanceId);
       LOGGER.error(errorMsg, e);
       throw new RuntimeException(errorMsg);
     }
   }
 
+  /**
+   * Add instance group tag for controller so that pinot controller can be assigned to lead controller resource.
+   */
+  private void addInstanceGroupTag() {
```

Comment:
Change this to `addTagIfNeeded` and only set instanceConfig if the tag does not exist already.

Reply:
Done.